File: pixiu/pkg/model/http.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package model

import (
	"time"
)

import (
	"github.com/mitchellh/mapstructure"
)

import (
	"github.com/apache/dubbo-go-pixiu/pixiu/pkg/logger"
)

// HttpConnectionManagerConfig
type HttpConnectionManagerConfig struct {
	RouteConfig       RouteConfiguration `yaml:"route_config" json:"route_config" mapstructure:"route_config"`
	HTTPFilters       []*HTTPFilter      `yaml:"http_filters" json:"http_filters" mapstructure:"http_filters"`
	ServerName        string             `yaml:"server_name" json:"server_name" mapstructure:"server_name"`
	IdleTimeoutStr    string             `yaml:"idle_timeout" json:"idle_timeout" mapstructure:"idle_timeout"`
	GenerateRequestID bool               `yaml:"generate_request_id" json:"generate_request_id" mapstructure:"generate_request_id"`
	TimeoutStr        string             `yaml:"timeout" json:"timeout" mapstructure:"timeout"`
	Timeout           time.Duration      `yaml:"-" json:"-" mapstructure:"-"`
}

// GRPCConnectionManagerConfig
type GRPCConnectionManagerConfig struct {
	RouteConfig RouteConfiguration `yaml:"route_config" json:"route_config" mapstructure:"route_config"`
	TimeoutStr  string             `yaml:"timeout" json:"timeout" mapstructure:"timeout"`
	Timeout     time.Duration      `yaml:"-" json:"-" mapstructure:"-"`
}

// DubboProxyConnectionManagerConfig
type DubboProxyConnectionManagerConfig struct {
	RouteConfig  RouteConfiguration `yaml:"route_config" json:"route_config" mapstructure:"route_config"`
	DubboFilters []*DubboFilter     `yaml:"dubbo_filters" json:"dubbo_filters" mapstructure:"dubbo_filters"`
	TimeoutStr   string             `yaml:"timeout" json:"timeout" mapstructure:"timeout"`
	Timeout      time.Duration      `yaml:"-" json:"-" mapstructure:"-"`
}

// HTTPFilter http filter
type HTTPFilter struct {
	Name   string                 `yaml:"name" json:"name" mapstructure:"name"`
	Config map[string]interface{} `yaml:"config" json:"config" mapstructure:"config"`
}

// DubboFilter dubbo filter
type DubboFilter struct {
	Name   string                 `yaml:"name" json:"name" mapstructure:"name"`
	Config map[string]interface{} `yaml:"config" json:"config" mapstructure:"config"`
}

type RequestMethod int32

const (
	METHOD_UNSPECIFIED = 0 + iota // (DEFAULT)
	GET
	HEAD
	POST
	PUT
	DELETE
	CONNECT
	OPTIONS
	TRACE
)

var RequestMethodName = map[int32]string{
	0: "METHOD_UNSPECIFIED",
	1: "GET",
	2: "HEAD",
	3: "POST",
	4: "PUT",
	5: "DELETE",
	6: "CONNECT",
	7: "OPTIONS",
	8: "TRACE",
}

var RequestMethodValue = map[string]int32{
	"METHOD_UNSPECIFIED": 0,
	"GET":                1,
	"HEAD":               2,
	"POST":               3,
	"PUT":                4,
	"DELETE":             5,
	"CONNECT":            6,
	"OPTIONS":            7,
	"TRACE":              8,
}

// HttpConfig the http config
type HttpConfig struct {
	IdleTimeoutStr  string `yaml:"idle_timeout" json:"idle_timeout" mapstructure:"idle_timeout"`
	ReadTimeoutStr  string `json:"read_timeout,omitempty" yaml:"read_timeout,omitempty" mapstructure:"read_timeout"`
	WriteTimeoutStr string `json:"write_timeout,omitempty" yaml:"write_timeout,omitempty" mapstructure:"write_timeout"`
	MaxHeaderBytes  int    `json:"max_header_bytes,omitempty" yaml:"max_header_bytes,omitempty" mapstructure:"max_header_bytes"`
}

func MapInStruct(cfg interface{}) *HttpConfig {
	var hc *HttpConfig
	if cfg != nil {
		if ok := mapstructure.Decode(cfg, &hc); ok != nil {
			logger.Error("Config error", ok)
		}
	}
	return hc
}

File: admin/web/src/services/http.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { API_BASE, API_STATUS } from '../app/constants'
import { clearSession, readSession, redirectToLogin } from './session'

export class ApiError extends Error {
  code: string
  raw: string
  issues: Array<{ path: string; code: string; message: string }>
  constructor(
    message: string,
    code = '',
    raw = '',
    issues: Array<{ path: string; code: string; message: string }> = [],
  ) {
    super(message)
    this.code = code
    this.raw = raw
    this.issues = issues
  }
}

function sessionHeaders() {
  const session = readSession()
  if (!session) return {}
  return { token: session.token, username: session.username }
}

function isAuthFailure(status: number, bodyData: unknown) {
  const body =
    bodyData && typeof bodyData === 'object' ? (bodyData as { code?: unknown; data?: unknown }) : {}
  if (status === 401 || status === 403) return true
  if (body.code !== API_STATUS.NOT_FOUND) return false
  const message = typeof body.data === 'string' ? body.data.toLowerCase() : ''
  return /token|login|access|authoriz|认证|登录|权限/.test(message)
}

function errorDetails(data: unknown) {
  if (!data || typeof data !== 'object' || Array.isArray(data))
    return { message: '', issues: [] as Array<{ path: string; code: string; message: string }> }
  const value = data as { message?: unknown; issues?: unknown }
  const issues = Array.isArray(value.issues)
    ? value.issues.filter(
        (issue): issue is { path: string; code: string; message: string } =>
          Boolean(issue) &&
          typeof issue === 'object' &&
          typeof (issue as { path?: unknown }).path === 'string' &&
          typeof (issue as { code?: unknown }).code === 'string' &&
          typeof (issue as { message?: unknown }).message === 'string',
      )
    : []
  return {
    message: typeof value.message === 'string' ? value.message : '',
    issues,
  }
}

function primitiveString(value: unknown, fallback = '') {
  if (typeof value === 'string') return value
  if (typeof value === 'number' || typeof value === 'boolean' || typeof value === 'bigint') {
    return value.toString()
  }
  return fallback
}

function responseDataString(value: unknown, fallback = '') {
  return primitiveString(value, fallback)
}

function httpErrorMessage(
  data: unknown,
  details: ReturnType<typeof errorDetails>,
  envelopeMessage: unknown,
  fallback: string,
) {
  if (typeof data === 'string') return data
  if (details.message) return details.message
  if (typeof envelopeMessage === 'string') return envelopeMessage
  return fallback
}

function codeErrorMessage(data: unknown, details: ReturnType<typeof errorDetails>, code: unknown) {
  if (typeof data === 'string') return data
  if (details.message) return details.message
  if (code === API_STATUS.NOT_FOUND) return '数据不存在'
  if (code === API_STATUS.CONCURRENT) return '操作冲突，请刷新后重试'
  return '请求失败'
}

export function isNotFoundError(error: unknown) {
  return error instanceof ApiError && error.code === API_STATUS.NOT_FOUND
}

export function parseArrayResponse<T>(data: unknown): T[] {
  let parsed = data
  if (typeof data === 'string') {
    try {
      parsed = JSON.parse(data) as unknown
    } catch (error) {
      throw new ApiError(
        '接口返回的数据不是有效 JSON',
        'BAD_RESPONSE',
        error instanceof Error ? error.message : primitiveString(error, '未知错误'),
      )
    }
  }
  if (!Array.isArray(parsed))
    throw new ApiError('接口返回的数据格式无效', 'BAD_RESPONSE', responseDataString(data))
  return parsed as T[]
}

export async function request<T>(path: string, init: RequestInit = {}) {
  const body = init.body
  const headers = new Headers(init.headers)
  Object.entries(sessionHeaders()).forEach(([key, value]) => headers.set(key, value))
  if (
    body &&
    !(body instanceof URLSearchParams) &&
    !(body instanceof FormData) &&
    !headers.has('Content-Type')
  )
    headers.set('Content-Type', 'application/json')
  const r = await fetch(API_BASE + path, { ...init, headers })
  const raw = await r.text()
  if (r.status === 401 || r.status === 403) {
    clearSession()
    redirectToLogin()
    throw new ApiError('登录状态已失效，请重新登录', 'AUTH_REQUIRED', raw)
  }
  if (r.status === 204) return undefined as T
  let bodyData: unknown
  try {
    bodyData = JSON.parse(raw)
  } catch {
    const boundary = raw.indexOf('}{')
    if (boundary > 0) {
      try {
        bodyData = JSON.parse(raw.slice(0, boundary + 1))
      } catch {
        bodyData = null
      }
    }
  }
  if (bodyData === null || typeof bodyData !== 'object')
    throw new ApiError(`接口返回了非 JSON 内容（HTTP ${r.status}）`, 'BAD_RESPONSE', raw)
  const envelope = bodyData as { code?: unknown; data?: unknown; message?: unknown }
  if (!r.ok) {
    if (isAuthFailure(r.status, envelope)) {
      clearSession()
      redirectToLogin()
      throw new ApiError(
        '登录状态已失效，请重新登录',
        'AUTH_REQUIRED',
        responseDataString(envelope.data, raw),
      )
    }
    const details = errorDetails(envelope.data)
    const message = httpErrorMessage(
      envelope.data,
      details,
      envelope.message,
      `请求失败（HTTP ${r.status}）`,
    )
    throw new ApiError(
      message,
      primitiveString(envelope.code, `HTTP_${r.status}`),
      raw,
      details.issues,
    )
  }
  if (envelope.code && envelope.code !== API_STATUS.SUCCESS) {
    if (isAuthFailure(r.status, envelope)) {
      clearSession()
      redirectToLogin()
      throw new ApiError(
        '登录状态已失效，请重新登录',
        'AUTH_REQUIRED',
        responseDataString(envelope.data),
      )
    }
    const details = errorDetails(envelope.data)
    const message = codeErrorMessage(envelope.data, details, envelope.code)
    throw new ApiError(
      message,
      primitiveString(envelope.code),
      responseDataString(envelope.data),
      details.issues,
    )
  }
  if (!('data' in envelope)) throw new ApiError('接口响应缺少 data 字段', 'BAD_RESPONSE', raw)
  return envelope.data as T
}
